cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,387 @@
+	}
+
+	yamlValues, err := yaml.JSONToYAML([]byte(jsonValues))
+	if err != nil {
+		l.logAndFatalf("jsonToYAML failed to parse values:\n%v\nError:\n%v", yamlValues, err)
+	}
+
+	return string(yamlValues)
+}
+
+func checkSupportedVersions(cur string, tar string) {
+	if cur == tar {
+		l.logAndFatalf("Abort. The current version %v equals to the target version %v.", cur, tar)
+	}
+
+	curMajor, curMinor, curPatch := parseVersionFormat(cur)
+	tarMajor, tarMinor, tarPatch := parseVersionFormat(tar)
+
+	if curMajor != tarMajor {
+		l.logAndFatalf("Abort. Major version upgrade is not supported: %v -> %v.", cur, tar)
+	}
+
+	if curMinor != tarMinor {
+		l.logAndFatalf("Abort. Minor version upgrade is not supported: %v -> %v.", cur, tar)
+	}
+
+	if curPatch == tarPatch {
+		l.logAndFatalf("Abort. The target version has been installed in the cluster.\n"+
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	if curPatch > tarPatch {
+		l.logAndFatalf("Abort. A newer version has been installed in the cluster.\n"+
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	if !supportedVersionMap[cur][tar] {
+		l.logAndFatalf("Abort. Upgrade is currently not supported: %v -> %v.", cur, tar)
+	}
+
+	l.logAndPrintf("Version check passed: %v -> %v.\n", cur, tar)
+}
+
+func parseVersionFormat(ver string) (int, int, int) {
+	fullVerArray := strings.Split(ver, "-")
+	if len(fullVerArray) == 0 {
+		l.logAndFatalf("Abort. Incorrect version: %v.", ver)
+	}
+	verArray := strings.Split(fullVerArray[0], ".")
+	if len(verArray) != 3 {
+		l.logAndFatalf("Abort. Incorrect version: %v.", ver)
+	}
+	major, err := strconv.Atoi(verArray[0])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect marjor version: %v.", verArray[0])
+	}
+	minor, err := strconv.Atoi(verArray[1])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect minor version: %v.", verArray[1])
+	}
+	patch, err := strconv.Atoi(verArray[2])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect patch version: %v.", verArray[2])
+	}
+	return major, minor, patch
+}
+
+func retrieveControlPlaneVersion(kubeClient kubernetes.ExecClient, istioNamespace string) string {
+	meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+	if e != nil {
+		l.logAndFatalf("Failed to retrieve Istio control plane version, error: %v", e)
+	}
+
+	if meshInfo == nil {
+		l.logAndFatalf("Istio control plane not found in namespace: %v", istioNamespace)
+	}
+
+	for _, remote := range *meshInfo {
+		l.logAndPrintf("Control Plane - %s pod - version: %s", remote.Component, remote.Info.Version)
+	}
+
+	return coalesceVersions(meshInfo)
+}
+
+func waitUpgradeComplete(kubeClient kubernetes.ExecClient, istioNamespace string, targetVer string) {
+	l.logAndPrintf("Start waiting for upgrade rollout to complete...")
+	sleepSeconds(20)
+	for i := 1; i <= 60; i++ {
+		if i != 1 {
+			sleepSeconds(10)
+		}
+		l.logAndPrintf("Waiting for upgrade rollout to complete, attempt #%v: ", i)
```

Comment:
It seems unnecessary to output this line since it doesn't tell the user much. I think it's sufficient to have dots appearing every second until the timeout is reached.

Reply:
Removed. Thanks so much!

Reply:
sorry, i meant the line above (waiting for upgrade...) as it's enough just to show the dots printing...